// @ng-toolkit/universal/esm2015/ngt-universal/document.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: ngt-universal/document.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
export class DocumentService {
    /**
     * @param {?} platformId
     */
    constructor(platformId) {
        if (!isPlatformBrowser(platformId)) {
            this._document = (/** @type {?} */ ((/** @type {?} */ ({
                getElementById: (/**
                 * @param {?} id
                 * @return {?}
                 */
                (id) => new HTMLElement()),
                getElementsByClassName: (/**
                 * @param {?} className
                 * @return {?}
                 */
                (className) => new HTMLCollection()),
                getElementsByName: (/**
                 * @param {?} name
                 * @return {?}
                 */
                (name) => new NodeList()),
                getElementsByTagName: (/**
                 * @param {?} name
                 * @return {?}
                 */
                (name) => new NodeList())
            }))));
        }
        else {
            console.log('in browser, assigning document');
            // this._document = document;
            this._document = (/** @type {?} */ ({}));
        }
    }
    /**
     * @return {?}
     */
    get nativeDocument() {
        console.log('get native document');
        console.log(this._document);
        return this._document;
    }
}
DocumentService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DocumentService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DocumentService.prototype._document;
}
