// @ng-toolkit/universal/esm2015/ngt-universal/local-storage-wrapper.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
export class LocalStorageWrapper {
    /**
     * @param {?} platformId
     */
    constructor(platformId) {
        if (isPlatformBrowser(platformId)) {
            this.localStorageRef = localStorage;
        }
        else {
            this.localStorageRef = {
                clear: () => {
                    /*noop*/ 
                },
                getItem: (key) => /** @type {?} */ (undefined),
                key: (index) => /** @type {?} */ (undefined),
                removeItem: (key) => {
                    /*noop*/ 
                },
                setItem: (key, value) => {
                    /*noop*/ 
                },
                length: 0
            };
        }
    }
    /**
     * @return {?}
     */
    get localStorage() {
        return this.localStorageRef;
    }
}
LocalStorageWrapper.decorators = [
    { type: Injectable },
];
/** @nocollapse */
LocalStorageWrapper.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
function LocalStorageWrapper_tsickle_Closure_declarations() {
    /** @type {?} */
    LocalStorageWrapper.prototype.localStorageRef;
}
