// @ng-toolkit/pwa/esm2015/ngt-pwa/sw-update-server.mock.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: ngt-pwa/sw-update-server.mock.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
export class SwUpdateServerMock {
    constructor() {
        this.available = new Subject();
        this.activated = new Subject();
        this.isEnabled = false;
    }
    /**
     * @return {?}
     */
    checkForUpdate() {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        (resolve) => resolve()));
    }
    /**
     * @return {?}
     */
    activateUpdate() {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        (resolve) => resolve()));
    }
}
if (false) {
    /** @type {?} */
    SwUpdateServerMock.prototype.available;
    /** @type {?} */
    SwUpdateServerMock.prototype.activated;
    /** @type {?} */
    SwUpdateServerMock.prototype.isEnabled;
}
