// @ng-toolkit/universal/esm5/ngt-universal/local-storage-wrapper.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
var LocalStorageWrapper = /** @class */ (function () {
    function LocalStorageWrapper(platformId) {
        if (isPlatformBrowser(platformId)) {
            this.localStorageRef = localStorage;
        }
        else {
            this.localStorageRef = {
                clear: function () {
                    /*noop*/ 
                },
                getItem: function (key) { return (undefined); },
                key: function (index) { return (undefined); },
                removeItem: function (key) {
                    /*noop*/ 
                },
                setItem: function (key, value) {
                    /*noop*/ 
                },
                length: 0
            };
        }
    }
    Object.defineProperty(LocalStorageWrapper.prototype, "localStorage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localStorageRef;
        },
        enumerable: true,
        configurable: true
    });
    LocalStorageWrapper.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    LocalStorageWrapper.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return LocalStorageWrapper;
}());
export { LocalStorageWrapper };
function LocalStorageWrapper_tsickle_Closure_declarations() {
    /** @type {?} */
    LocalStorageWrapper.prototype.localStorageRef;
}
