// @ng-toolkit/universal/esm5/ngt-universal/local-storage-wrapper.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: ngt-universal/local-storage-wrapper.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
var LocalStorageWrapper = /** @class */ (function () {
    function LocalStorageWrapper(platformId) {
        if (isPlatformBrowser(platformId)) {
            this.localStorageRef = localStorage;
        }
        else {
            this.localStorageRef = {
                clear: (/**
                 * @return {?}
                 */
                function () { }),
                getItem: (/**
                 * @param {?} key
                 * @return {?}
                 */
                function (key) { return (/** @type {?} */ (undefined)); }),
                key: (/**
                 * @param {?} index
                 * @return {?}
                 */
                function (index) { return (/** @type {?} */ (undefined)); }),
                removeItem: (/**
                 * @param {?} key
                 * @return {?}
                 */
                function (key) { }),
                setItem: (/**
                 * @param {?} key
                 * @param {?} value
                 * @return {?}
                 */
                function (key, value) { }),
                length: 0
            };
        }
    }
    Object.defineProperty(LocalStorageWrapper.prototype, "localStorage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localStorageRef;
        },
        enumerable: true,
        configurable: true
    });
    LocalStorageWrapper.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LocalStorageWrapper.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return LocalStorageWrapper;
}());
export { LocalStorageWrapper };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LocalStorageWrapper.prototype.localStorageRef;
}
