// @ng-toolkit/universal/esm5/ngt-universal/ngt-universal.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: ngt-universal/ngt-universal.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { WindowService } from './window.service';
import { LocalStorageWrapper } from './local-storage-wrapper.service';
import { DocumentService } from './document.service';
export { DocumentService } from './document.service';
/** @type {?} */
export var WINDOW = new InjectionToken('ng-toolkit-window');
/**
 * @param {?} windowService
 * @return {?}
 */
export function windowFactory(windowService) {
    return windowService.nativeWindow;
}
/** @type {?} */
export var LOCAL_STORAGE = new InjectionToken('ng-toolkit-local-storage');
/**
 * @param {?} localStorageWrapper
 * @return {?}
 */
export function localStorageFactory(localStorageWrapper) {
    return localStorageWrapper.localStorage;
}
/** @type {?} */
export var NGT_DOCUMENT = new InjectionToken('ng-toolkit-document');
/**
 * @param {?} documentService
 * @return {?}
 */
export function documentFactory(documentService) {
    console.log('document factory');
    return documentService.nativeDocument;
}
var NgtUniversalModule = /** @class */ (function () {
    function NgtUniversalModule() {
    }
    NgtUniversalModule.decorators = [
        { type: NgModule, args: [{
                    providers: [
                        WindowService,
                        { provide: WINDOW, useFactory: windowFactory, deps: [WindowService] },
                        DocumentService,
                        { provide: NGT_DOCUMENT, useFactory: documentFactory, deps: [DocumentService] },
                        LocalStorageWrapper,
                        { provide: LOCAL_STORAGE, useFactory: localStorageFactory, deps: [LocalStorageWrapper] }
                    ]
                },] }
    ];
    return NgtUniversalModule;
}());
export { NgtUniversalModule };
