// @ng-toolkit/pwa/esm5/ngt-pwa/sw-update-server.mock.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: ngt-pwa/sw-update-server.mock.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
var SwUpdateServerMock = /** @class */ (function () {
    function SwUpdateServerMock() {
        this.available = new Subject();
        this.activated = new Subject();
        this.isEnabled = false;
    }
    /**
     * @return {?}
     */
    SwUpdateServerMock.prototype.checkForUpdate = /**
     * @return {?}
     */
    function () {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) { return resolve(); }));
    };
    /**
     * @return {?}
     */
    SwUpdateServerMock.prototype.activateUpdate = /**
     * @return {?}
     */
    function () {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) { return resolve(); }));
    };
    return SwUpdateServerMock;
}());
export { SwUpdateServerMock };
if (false) {
    /** @type {?} */
    SwUpdateServerMock.prototype.available;
    /** @type {?} */
    SwUpdateServerMock.prototype.activated;
    /** @type {?} */
    SwUpdateServerMock.prototype.isEnabled;
}
